File: 073_Construct Binary Tree from Inorder and Postorder Traversal.cpp
// Construct Binary Tree from Inorder and Postorder Traversal
// Build a binary tree given its inorder and postorder traversal arrays.
// Solution: use hashmap for inorder indices and recursive builder using postorder index from end.
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {int val; TreeNode* left; TreeNode* right; TreeNode(int x): val(x), left(nullptr), right(nullptr) {}};

class Solution {
    unordered_map<int,int> idx;
    vector<int> post;
    int postIndex;
public:
    TreeNode* build(int l, int r) {
        if (l>r) return nullptr;
        int rootVal = post[postIndex--];
        TreeNode* root = new TreeNode(rootVal);
        int mid = idx[rootVal];
        // build right first because postorder gives root after children
        root->right = build(mid+1, r);
        root->left = build(l, mid-1);
        return root;
    }
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        post = postorder;
        postIndex = (int)post.size()-1;
        idx.clear();
        for (int i=0;i<inorder.size();++i) idx[inorder[i]] = i;
        return build(0, (int)inorder.size()-1);
    }
};

void printInorder(TreeNode* r) {
    if (!r) return;
    printInorder(r->left);
    cout<<r->val<<" ";
    printInorder(r->right);
}

int main(){
    Solution sol;
    vector<int> inorder = {9,3,15,20,7};
    vector<int> postorder = {9,15,7,20,3};
    TreeNode* root = sol.buildTree(inorder, postorder);
    cout<<"Inorder of built tree: "; printInorder(root); cout<<"\n";
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(n)

File: 073_Construct Binary Tree from Inorder and Postorder Traversal.py
"""
Construct Binary Tree from Inorder and Postorder Traversal
Build a binary tree given inorder and postorder traversal arrays.
"""
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        idx = {v:i for i,v in enumerate(inorder)}
        self.postIndex = len(postorder)-1

        def build(l, r):
            if l>r: return None
            rootVal = postorder[self.postIndex]; self.postIndex -= 1
            root = TreeNode(rootVal)
            mid = idx[rootVal]
            root.right = build(mid+1, r)
            root.left = build(l, mid-1)
            return root

        return build(0, len(inorder)-1)

def inorder_traversal(root):
    return inorder_traversal(root.left) + [root.val] + inorder_traversal(root.right) if root else []

if __name__ == '__main__':
    inorder = [9,3,15,20,7]
    postorder = [9,15,7,20,3]
    root = Solution().buildTree(inorder, postorder)
    print('Inorder of built tree:', inorder_traversal(root))

# Time Complexity: O(n)
# Space Complexity: O(n)
